google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -319,52 +336,359 @@ private boolean isAlive() {
+    // This callback handles retries, and sets message futures
+
+    // Check if ack or nack
+    boolean setResponseOnSuccess = (!isModack || (deadlineExtensionSeconds == 0)) ? true : false;
+
+    return new ApiFutureCallback<Empty>() {
+      @Override
+      public void onSuccess(Empty empty) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+        for (AckRequestData ackRequestData : ackRequestDataList) {
+          // This will check if a response is needed, and if it has already been set
+          ackRequestData.setResponse(AckResponse.SUCCESSFUL, setResponseOnSuccess);
+          // Remove from our pending operations
+          pendingRequests.remove(ackRequestData);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        // Remove from our pending operations
+        ackOperationsWaiter.incrementPendingCount(-1);
+        List<AckRequestData> ackRequestDataArrayRetryList = new ArrayList<>();
+        try {
+          Map<String, String> metadataMap = getMetadataMapFromThrowable(t);
+          ackRequestDataList.forEach(
+              ackRequestData -> {
+                String ackId = ackRequestData.getAckId();
+                if (metadataMap.containsKey(ackId)) {
+                  // An error occured
+                  String errorMessage = metadataMap.get(ackId);
+                  if (errorMessage.startsWith(TRANSIENT_FAILURE_METADATA_PREFIX)) {
+                    // Retry all "TRANSIENT_*" error messages - do not set message future
+                    logger.log(Level.WARNING, "Transient error message, will resend", errorMessage);
+                    ackRequestDataArrayRetryList.add(ackRequestData);
+                  } else if (errorMessage.equals(PERMANENT_FAILURE_INVALID_ACK_ID_METADATA)) {
+                    // Permanent failure, send
+                    logger.log(
+                        Level.WARNING,
+                        "Permanent error invalid ack id message, will not resend",
+                        errorMessage);
+                    ackRequestData.setResponse(AckResponse.INVALID, setResponseOnSuccess);
+                  } else {
+                    logger.log(
+                        Level.WARNING, "Unknown error message, will not resend", errorMessage);
+                    ackRequestData.setResponse(AckResponse.OTHER, setResponseOnSuccess);
+                  }
+                } else {
+                  ackRequestData.setResponse(AckResponse.SUCCESSFUL, setResponseOnSuccess);
+                }
+                // Remove from our pending
+                pendingRequests.remove(ackRequestData);
+              });
+        } catch (InvalidProtocolBufferException e) {
+          // If we fail to parse out the errorInfo, we should retry all
+          logger.log(
+              Level.WARNING, "Exception occurred when parsing throwable {0} for errorInfo", t);
+          ackRequestDataArrayRetryList.addAll(ackRequestDataArrayRetryList);
```

Comment:
I think this should be ackRequestDataArrayRetryList.addAll(ackRequestDataList) right?

Reply:
Yes, fixed.